src/apple/alohalytics_objc.mm

```diff
@@ -0,0 +1,354 @@
+    return ToStdString(object);
+  } else if ([object isKindOfClass:[NSObject class]]) {
+    return ToStdString(((NSObject *)object).description);
+  }
+  return "ERROR: Trying to log neither NSString nor NSObject-inherited object.";
+}
+
+// Safe conversion from [possible nil] NSDictionary.
+static TStringMap ToStringMap(NSDictionary * nsDictionary) {
+  TStringMap map;
+  for (NSString * key in nsDictionary) {
+    map[ToStdString(key)] = ToStdStringSafe([nsDictionary objectForKey:key]);
+  }
+  return map;
+}
+
+// Safe conversion from [possible nil] NSArray.
+static TStringMap ToStringMap(NSArray * nsArray) {
+  TStringMap map;
+  std::string key;
+  for (id item in nsArray) {
+    if (key.empty()) {
+      key = ToStdStringSafe(item);
+      map[key] = "";
+    } else {
+      map[key] = ToStdStringSafe(item);
+      key.clear();
+    }
+  }
+  return map;
+}
+
+// Safe extraction from [possible nil] CLLocation to alohalytics::Location.
+static Location ExtractLocation(CLLocation * l) {
+  Location extracted;
+  if (l) {
+    if (l.horizontalAccuracy >= 0) {
+      extracted.SetLatLon([l.timestamp timeIntervalSince1970] * 1000.,
+                           l.coordinate.latitude, l.coordinate.longitude,
+                           l.horizontalAccuracy);
+    }
+    if (l.verticalAccuracy >= 0) {
+      extracted.SetAltitude(l.altitude, l.verticalAccuracy);
+    }
+    if (l.speed >= 0) {
+      extracted.SetSpeed(l.speed);
+    }
+    if (l.course >= 0) {
+      extracted.SetBearing(l.course);
+    }
+    // We don't know location source on iOS.
+  }
+  return extracted;
+}
+
+// uint64_t timestamp of Documents folder modification date in millis from 1970, represented as a string.
+// Can be interpreted as a "first app launch time" or an approx. "app install time".
+static std::string DocumentsTimestampMillis() {
+  NSArray * paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
+  NSDictionary * attributes = [[NSFileManager defaultManager] attributesOfItemAtPath:[paths firstObject] error:nil];
+  if (attributes) {
+    NSDate * date = [attributes objectForKey:NSFileModificationDate];
+    return std::to_string(static_cast<uint64_t>([date timeIntervalSince1970] * 1000.));
+  }
+  return std::string("0");
+}
+
+// uint64_t timestamp of app bundle folder modification date in millis from 1970, represented as a string.
+// Can be interpreted as an "app update time".
+static std::string BundleTimestampMillis() {
+  NSDictionary * attributes = [[NSFileManager defaultManager] attributesOfItemAtPath:[[NSBundle mainBundle] executablePath] error:nil];
+  if (attributes) {
+    NSDate * date = [attributes objectForKey:NSFileModificationDate];
+    return std::to_string(static_cast<uint64_t>([date timeIntervalSince1970] * 1000.));
+  }
+  return std::string("0");
+}
+
+static std::string RectToString(CGRect const & rect) {
+  return std::to_string(rect.origin.x) + " " + std::to_string(rect.origin.y) + " "
```

Comment:
High-level comment: You seem to do a lot of string packing here. Which brings us back to the question: Why not JSON?